plutus-core/plutus-ir/Language/PlutusIR/Parser.hs

```diff
@@ -209,38 +157,90 @@ version = lexeme $ do
     void $ char '.'
     PLC.Version p x y <$> Lex.decimal
 
-constant :: Parser (PLC.Some (PLC.ValueOf PLC.DefaultUni))
-constant =
-    (reservedWord "integer" >> PLC.someValue <$> integer) <|> (reservedWord "bytestring" >> PLC.someValue <$> bytestring)
-
 recursivity :: Parser Recursivity
 recursivity = inParens $ (reservedWord "rec" >> return Rec) <|> (reservedWord "nonrec" >> return NonRec)
 
 strictness :: Parser Strictness
 strictness = inParens $ (reservedWord "strict" >> return Strict) <|> (reservedWord "nonstrict" >> return NonStrict)
 
-funType :: Parser (Type TyName PLC.DefaultUni SourcePos)
+funType
+    :: PLC.Parsable (PLC.Some uni)
+    => Parser (Type TyName uni SourcePos)
 funType = TyFun <$> reservedWord "fun" <*> typ <*> typ
 
-allType :: Parser (Type TyName PLC.DefaultUni SourcePos)
+allType
+    :: PLC.Parsable (PLC.Some uni)
+    => Parser (Type TyName uni SourcePos)
 allType = TyForall <$> reservedWord "all" <*> tyVar <*> kind <*> typ
 
-lamType :: Parser (Type TyName PLC.DefaultUni SourcePos)
+lamType
+    :: PLC.Parsable (PLC.Some uni)
+    => Parser (Type TyName uni SourcePos)
 lamType = TyLam <$> reservedWord "lam" <*> tyVar <*> kind <*> typ
 
-ifixType :: Parser (Type TyName PLC.DefaultUni SourcePos)
+ifixType
+    :: PLC.Parsable (PLC.Some uni)
+    => Parser (Type TyName uni SourcePos)
 ifixType = TyIFix <$> reservedWord "ifix" <*> typ <*> typ
 
-conType :: Parser (Type TyName PLC.DefaultUni SourcePos)
+conType
+    :: PLC.Parsable (PLC.Some uni)
+    => Parser (Type TyName uni SourcePos)
 conType = reservedWord "con" >> builtinType
 
-builtinType :: Parser (Type TyName PLC.DefaultUni SourcePos)
+-- | Turn a parser that can succeed without consuming any input into one that fails in this case.
+enforce :: Parser a -> Parser a
+enforce p = do
+    (input, x) <- match p
+    guard . not $ T.null input
+    pure x
+
+-- | Consume a chunk of text and either stop on whitespace (and consume it) or on a \')\' (without
+-- consuming it). We need this for collecting a @Text@ to pass it to 'PLC.parse' in order to
+-- parse a built-in type or a constant. This is neither efficient (because of @manyTill anySingle@),
+-- nor future-proof (what if some future built-in type has parens in its syntax?). A good way of
+-- resolving this would be to turn 'PLC.parse' into a proper parser rather than just a function
+-- from @Text@, but PLC's parser also uses 'PLC.parse' and it's currently not @megaparsec@-based,
+-- and so mixing two distinct parsing machines doesn't sound too exciting.
+closedChunk :: Parser T.Text
+closedChunk = T.pack <$> manyTill anySingle end where
+    end = enforce whitespace <|> void (lookAhead $ char ')')
```

Comment:
I think this may have problems with things like `(con string "yes (no)") and perhaps also the unit value `(con ())` (although maybe we shouldn't have that).  I think it should be possible to use the PLC Lexer to tokenise the input to the PIR parser: that would solve that problem, but would require reworking this parser quite a bit.  Also we'd have to extend `Lexer.x` and I'm not sure if that would conflict with the PLC syntax.  Is the PIR syntax (or just the PUR extensions) written down anywhere?  You can work it out from the parser, but a separate grammar would be easier to understand.

Reply:
> I think this may have problems with things like `(con string "yes (no)")` and perhaps also the unit value `(con ())` (although maybe we shouldn't have that). 

Rrright... Thank you. That clearly needs to be reworked, but I'm not quite sure how... And there can be expressions like `(con string "yes :)")`, so we can't just look for matching parens for example.

I think what we should do is make `Parsable` `megaparsec`-based. Then constant parsers consume as many symbols as they like and we won't need to try to predict in the main parser what is a constant and what is formatting.

Anyway, the previous PIR parser was only able to parse `integer` and `bytestring`, so at least I'm not making it worse.

> Is the PIR syntax (or just the PIR extensions) written down anywhere? You can work it out from the parser, but a separate grammar would be easier to understand.

Dunno. Maybe in the original Michael's proposal? Don't even remember where that was.

Reply:
If you look in `Lexer.x` there's extensive commentary about how it parses constants.  We could steal the relevant regular expressions from there or reuse the entire lexer.  I'm not convinced that letting types do their own lexical analysis is a good idea: doing these things correctly can be quite tricky and if you got it wrong it might mess up the main parser in some subtle way that doesn't show up until you encounter an unexpected situation.  The way that it's done at the moment provides reasonably general syntax for constants and I'm fairly confident it's correct.  I guess I feel that this is low-level stuff that should be done once and then shut away in a box where it won't cause any trouble.

Reply:
Ok, that makes sense. Wanna volunteer to do that? Not in this PR, I guess.

Reply:
Although perhaps it's really low priority.

Reply:
OK.  I've been thinking for a while that the parsers should be consistent but not doing anything about it.  I'll open a JIRA issue.  As you say, I don't think it's massively ugent.